lona/client/_lona/rendering-engine.js

```diff
@@ -0,0 +1,637 @@
+        let widget_data = this._widget_data.get(node_id);
+        let new_data = undefined;
+
+        key_path.forEach(key => {
+            parent_data = widget_data;
+            widget_data = widget_data[key];
+        });
+
+        // SET
+        if(operation == Lona.protocol.OPERATION.SET) {
+            widget_data[data[0]] = data[1];
+
+        // RESET
+        } else if(operation == Lona.protocol.OPERATION.RESET) {
+            if(parent_data === undefined) {
+                this._widget_data.set(node_id, data[0]);
+
+            } else {
+                parent_data = data[0];
+
+            };
+
+        // CLEAR
+        } else if(operation == Lona.protocol.OPERATION.CLEAR) {
+            if(widget_data instanceof Array) {
+                new_data = [];
+
+            } else if(widget_data instanceof Object) {
+                new_data = {};
+
+            };
+
+            if(parent_data === undefined) {
+                this._widget_data.set(node_id, new_data);
+
+            } else {
+                parent_data[key_path[key_path.length-1]] = new_data;
+
+            };
+
+        // INSERT
+        } else if(operation == Lona.protocol.OPERATION.INSERT) {
+            widget_data.splice(data[0], 0, data[1]);
+
+        // REMOVE
+        } else if(operation == Lona.protocol.OPERATION.REMOVE) {
+            if(widget_data instanceof Array) {
+                widget_data.splice(data[0], 1);
+
+            } else if(data instanceof Object) {
+                delete widget_data[data[0]];
+
+            };
+        };
+
+       this._widgets_to_update.splice(0, 0, node_id);
+    };
+
+    _apply_patch(patch) {
+        const patch_type = patch[1];
+
+        if(patch_type == Lona.protocol.PATCH_TYPE.WIDGET_DATA) {
+            this._apply_patch_to_widget_data(patch);
+
+        } else if(patch_type == Lona.protocol.PATCH_TYPE.NODES) {
+            this._apply_patch_to_child_nodes(patch);
+
+        } else {
+            this._apply_patch_to_node(patch);
+        };
+    };
+
+    // public api -------------------------------------------------------------
+    show_html(html) {
+        const message_type = html[0];
+        const data = html[1];
+
+        // HTML
+        if (message_type == Lona.protocol.DATA_TYPE.HTML) {
+            this._root.innerHTML = data;
+            this._clear_node_cache();
+            this.lona_window._input_event_handler.patch_input_events_recursively(this._root);
+
+        // HTML tree
+        } else if(message_type == Lona.protocol.DATA_TYPE.HTML_TREE) {
+            this._clear_node_cache();
+
+            const node = this._render_node(data)
+
+            this._clear();
+            this._root.appendChild(node);
+
+        // HTML update
+        } else if(message_type == Lona.protocol.DATA_TYPE.HTML_UPDATE) {
+            this._widgets_to_setup = [];
+            this._widgets_to_update = [];
+
+            data.forEach(patch => {
+                this._apply_patch(patch);
+            });
```

Comment:
nit:
```suggestion
            data.forEach(this._apply_patch);
```

Reply:
This won't work. The callback given to `data.forEach` has to be an arrow-function to preserve the scope of `this`. If written like this, this exception raises:

```
TypeError: Cannot read properties of undefined (reading '_apply_patch_to_node')                                                                          
    at _apply_patch (http://localhost:8080/static//_lona/rendering-engine.js:599:18)                                                                     
    at Array.forEach (<anonymous>)                                                                                                                       
    at LonaRenderingEngine.show_html (http://localhost:8080/static//_lona/rendering-engine.js:628:18)                                                    
    at LonaWindow._handle_websocket_message (http://localhost:8080/static//_lona/window.js:195:40)                                                       
    at LonaWindow.handle_websocket_message (http://localhost:8080/static//_lona/window.js:215:25)                                                        
    at WebSocket._handle_raw_websocket_message (http://localhost:8080/static//_lona/context.js:235:25) 
```

Reply:
Wow, didn't know. Thanks!